### ISSUE TITLE
How to mount an SD card or USB OTG on udroid?

### ISSUE THREAD

**Issue by hansm629** [RandomCoderOrg/ubuntu-on-android#236]
Is there a way to mount the SD card or USB OTG on the host device in an udroid environment?
If possible, what settings should I make in Termux?

installing and configuring termux-usb on termux?

The termux-usb setup method I found through Googling was difficult T_T

**Comment by ThatMG393** [comment]
https://github.com/termux/termux-packages/issues/7037

**Comment by vinci-d** [comment]
https://udroid-rc.gitbook.io/udroid-wiki/udroid-landing/tips-and-tricks/external-sdcard-access


Just need to find out the mount point of OTG with `df -h` ( in **Mounted on** colomn) matching with your OTG storage.

> Not guaranteed to work in every device tho.
>
> But this is the easiest one

**Comment by hansm629** [comment]
> https://udroid-rc.gitbook.io/udroid-wiki/udroid-landing/tips-and-tricks/external-sdcard-access
> 
> Just need to find out the mount point of OTG with `df -h` ( in **Mounted on** colomn) matching with your OTG storage.
> 
> > Not guaranteed to work in every device tho.
> > But this is the easiest one

@vinci-d 

Thank you for answer :)
I have a few additional questions.

Q1. Is this possible in a without root environment?


Q2. When setting storage read/write permissions using adb after connecting a host device to a PC, do I need to do so while the SD card or USB OTG is connected?
```
adb shell pm grant com.termux android.permission.READ_EXTERNAL_STORAGE
adb shell pm grant com.termux android.permission.WRITE_EXTERNAL_STORAGE
``` 

Q3. Is it correct to set the above permissions with adb and enter the command below in Termux?
```
$ df | grep /storage | cut -d "/" -f 4-
``` 


Q4. After logging in to Udroid, do I need to enter a separate command for mounting?

**Comment by SaicharanKandukuri** [comment]
> https://udroid-rc.gitbook.io/udroid-wiki/udroid-landing/tips-and-tricks/external-sdcard-access
> 
> Just need to find out the mount point of OTG with `df -h` ( in **Mounted on** colomn) matching with your OTG storage.
> 
> > Not guaranteed to work in every device tho.
> > But this is the easiest one

This trick doesn't work if OTG or Sdcard is not handled by [`fuse` (A userspace filesystem mounted by a non-privileged (non-root) user)](https://source.android.com/docs/core/storage/fuse-passthrough)
apparently, in Samsung devices at least, OTG and sd card are mounted with `fuse` which made it accessible in termux
but in AOSP 11+ and pixel ROM, only sd card is mounted with fuse

---
@hansm629 

To access the SDcard, you can directly find the mount point with `df -h` and `cd` to the mount point

<img src="https://github.com/RandomCoderOrg/ubuntu-on-android/assets/68287637/9f98ec49-a5e9-4c2c-8fd9-a61c4a4ce9b8"  height="600"/>

> check out this screenshot.  I found the mount of sd card ( external storage ) and made a small file change
> All required are set in the termux app (  https://github.com/termux/termux-app/blob/c84d4804c82c8325d1037fc20a5a847026fa11ca/app/src/main/AndroidManifest.xml#L25 )

---

> PS: ill update the documentation in wiki

**Comment by elvisisvan** [comment]
> > https://udroid-rc.gitbook.io/udroid-wiki/udroid-landing/tips-and-tricks/external-sdcard-access
> > Just need to find out the mount point of OTG with `df -h` ( in **Mounted on** colomn) matching with your OTG storage.
> > > Not guaranteed to work in every device tho.
> > > But this is the easiest one
> 
> This trick doesn't work if OTG or Sdcard is not handled by [`fuse` (A userspace filesystem mounted by a non-privileged (non-root) user)](https://source.android.com/docs/core/storage/fuse-passthrough) apparently, in Samsung devices at least, OTG and sd card are mounted with `fuse` which made it accessible in termux but in AOSP 11+ and pixel ROM, only sd card is mounted with fuse
> 
> @hansm629
> 
> To access the SDcard, you can directly find the mount point with `df -h` and `cd` to the mount point
> 
> <img alt="" height="600" src="https://private-user-images.githubusercontent.com/68287637/279855925-9f98ec49-a5e9-4c2c-8fd9-a61c4a4ce9b8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI1ODQ3MjAsIm5iZiI6MTcwMjU4NDQyMCwicGF0aCI6Ii82ODI4NzYzNy8yNzk4NTU5MjUtOWY5OGVjNDktYTVlOS00YzJjLThmZDktYTYxYzRhNGNlOWI4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE0VDIwMDcwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRiZjM4NWRjZWJhMmU1ZTEzMjM4ZjE5NzVmZTY5NTdjMWVhYmExNjc4NTJjNGQxYmQ1MTU0YjRhZTQwZTI0MDMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.tp2bEBaGpTTAiggRYg9d4AKWUkubYVpStmi5IyyU2gM">
> > check out this screenshot.  I found the mount of sd card ( external storage ) and made a small file change
> > All required are set in the termux app (  https://github.com/termux/termux-app/blob/c84d4804c82c8325d1037fc20a5a847026fa11ca/app/src/main/AndroidManifest.xml#L25 )
> 
> > PS: ill update the documentation in wiki

how is it that EASY on your side?!? it's literally plug-and-play! 
meanwhile i've been jumping around multiple identical issues (and went a whole loop between them) hoping to simply be able to navigate into external storage like you just did;
i went from editing /system/etc/permissions/platform.xml 
![Screenshot_20231215-024930](https://github.com/RandomCoderOrg/ubuntu-on-android/assets/69122770/408f4af9-b966-429a-be60-0cd2ebc07239)

to (try) allowing MANAGE_EXTERNAL_STORAGE permission for termux (i did so via App Ops & Shizuku, Pixel 6a, android 13, rooted) 
![Screenshot_20231215-030635](https://github.com/RandomCoderOrg/ubuntu-on-android/assets/69122770/7149f19c-f26a-4fdf-a570-ba0736736e50)

to running `df -h` command exactly as you did but the output does not include what is expected like your 77FF-BE6D device, according to MiXplorer mine is 2116-31F4 but it does  not appear under /storage/emulated line like yours
![Screenshot_20231215-030055](https://github.com/RandomCoderOrg/ubuntu-on-android/assets/69122770/9c80181d-9497-43c8-a48a-400578be32e0)

may i ask what kind of device is your external storage? and its file system type?

mine is a kingston ssd placed in an enclosure connecting via type-c to type-c cable, extFS file system, mounted using exFAT/NTFS for USB software

**Comment by AlainKnaff** [comment]
I found that the following command, executed as user shell (via adb shell over USB cable) sets up the device up so that both secondary SD card and USB sticks are mounted under /storage with fuse:

sm set-force-adoptable on

After this, these devices can be accessed from Termux. /storage directory is unreadable, but names are displayed just fine with df.

Setting only applies to devices inserted after command has been issued. Setting is remembered across reboots

Tested with stock Android 14 on a Pixel 7a, and with LineageOS 20 on a Galaxy Tab S3.
